### DISCUSSION TITLE
libbsd project homepage says it is supported on macOS, but homebrew says it requires linux

### DISCUSSION THREAD

**Discussion by bimimicah** [Homebrew/discussions#5755] [Writing Formulae/Casks] [answered]
### Description of issue

The [libbsd project homepage](https://libbsd.freedesktop.org/wiki/) says it is supported "out of the box" on macOS, but the [libbsd homebrew formula](https://github.com/Homebrew/homebrew-core/blob/f5f4dcae6bca5579ac5883e829aa6c9429bd29b2/Formula/lib/libbsd.rb) says it requires linux. Presumably just removing that line and rebuilding would solve this, if it is truly "out of the box"?

Is this something that can be easily fixed? I don't have access to a mac machine, so I cannot work on this myself.

**Comment by Bo98** [comment]
It was imported from the linuxbrew-core so is probably simply the case that nobody has requested or needed it yet. This is probably because macOS is a BSD-like OS so its feature set largely already implemented in the OS itself, making libbsd usually unnecessary. Is there a particular feature you're looking for? Would help us understand the use case better.

**Reply by bimimicah** [reply]
I am a maintainer of mod_authz_unixgroup, an authorization module for Apache webserver (httpd). A [recent pull request to fix compatibility issues with modern unix auth services](https://github.com/phokz/mod-auth-external/pull/55) required us to switch from the posix-standard <grp.h> to the libbsd version (we now use getgrouplist() and gid_from_group()).
As far as I can tell from our [GitHub Actions build errors](https://github.com/phokz/mod-auth-external/actions/runs/11976148222/job/33391175229), getgrouplist() has a different function declaration and gid_from_group() does not exist at all on macOS.

**Reply by bimimicah** [reply]
(To prevent any confusion, that particular build run linked above used the standard grp.h header include and did not attempt to compile against libbsd. The current PR content may differ.)

**Reply by Bo98** [reply]
`getgrouplist` does indeed have slightly different types which is a common annoyance though is ultimately still compatible with a cast (https://github.com/apple-oss-distributions/Libc/blob/af11da5ca9d527ea2f48bb7efbd0f0f2a4ea4812/sys/getgroups.c#L53 which can be wrapped in `#ifdef __APPLE__`). Worth noting that on Linux, glibc also has `getgrouplist` without needing `libbsd` if you use `_DEFAULT_SOURCE`/`_BSD_SOURCE`.

`gid_from_group` is indeed not available on macOS. `getgrnam`/`getgrnam_r` is the portable way to do that.

(To be clear: I'd still be OK enabling libbsd on macOS if someone made a pull request for that)

**Reply by bimimicah** [reply]
> (To be clear: I'd still be OK enabling libbsd on macOS if someone made a pull request for that)

I'd be happy to make a PR, removing that requirement line from the formula, but since I can't really test it, that seems a bit dodgy, right?

I'll go ahead and try it, and we can see if it passes build tests, for starters.

For our project, I would prefer to keep everything as standard as possible, sticking to APIs that are cross-platform as much as possible: APR (Apache Portable Runtime), POSIX, libbsd. I'd rather not have to add platform-specific code, unless we absolutely have to (Windows....)

**Reply by bimimicah** [reply]
> `gid_from_group` is indeed not available on macOS. `getgrnam`/`getgrnam_r` is the portable way to do that.

I'll have to go back and recheck with the PR author. We were specifically removing our use of getgrnam() to solve the compatibility issues with sssd.

**Reply by Bo98** [reply]
> > `gid_from_group` is indeed not available on macOS. `getgrnam`/`getgrnam_r` is the portable way to do that.
> 
> I'll have to go back and recheck with the PR author. We were specifically removing our use of getgrnam() to solve the compatibility issues with sssd.

Interesting given `gid_from_group` in BSDs is literally just a `getgrnam_r` + a cache. There shouldn't be any functional difference beyond caching.

**Reply by Bo98** [reply]
> I'd be happy to make a PR, removing that requirement line from the formula, but since I can't really test it, that seems a bit dodgy, right?

CI will do a build check + the simple test that you see in the formula itself. It's probably good enough. I'm away for the next week so am unable to much else myself.

**Reply by bimimicah** [reply]
> Interesting given `gid_from_group` in BSDs is literally just a `getgrnam_r` + a cache. There shouldn't be any functional difference beyond caching.

<s>I believe that's the whole issue, actually.</s> That for some auth providers (`sssd` in this case; LDAP/AD/etc.), the cached data is incomplete. The way it was described to me, `grp->gr_mem` may just be completely empty. Which makes sense if you don't want to copy down the whole LDAP directory to your local computer (or may not even have permission to do so, depending on the logged-in user), I suppose.

EDIT: I misread your message and missed that you were talking about `gid_from_group`'s cache, sorry. If that is the case then we probably can make the change. The rest of this message I'll leave as background info.

**Reply by bimimicah** [reply]
I will have to ask the commit author to test to know for sure, but I think the PR's change to use `getgrouplist()` solves that issue, and we may be able to use `getgrnam()` in this particular case where we only need the `grp->gr_gid` field

**Reply by bimimicah** [reply]
> > I'd be happy to make a PR, removing that requirement line from the formula, but since I can't really test it, that seems a bit dodgy, right?
> 
> CI will do a build check + the simple test that you see in the formula itself. It's probably good enough. I'm away for the next week so am unable to much else myself.

A draft PR for CI tests was created here [#198648](https://github.com/Homebrew/homebrew-core/pull/198648). If that passes all the checks, then I will read all the guidelines and make a proper PR. I hope that is acceptable.

**Reply by bimimicah** [reply]
> > I'd be happy to make a PR, removing that requirement line from the formula, but since I can't really test it, that seems a bit dodgy, right?
> 
> CI will do a build check + the simple test that you see in the formula itself. It's probably good enough. I'm away for the next week so am unable to much else myself.

I think I have everything in order for the final PR. Please let me know if there is anything else you need from me. Thanks for your help and advice on libbsd alternatives.

**Comment by bimimicah** [comment, marked as answer]
This was fixed in PR [#198648](https://github.com/Homebrew/homebrew-core/pull/198648). Thanks for everyone's help.
